### PR TITLE
Introduce edm::Async service, and use it in CUDA and Alpaka modules

### LINKED ISSUE #916: Open a PR and address review comments

### LINKED ISSUE #29188: Consider edm::async()
PR https://github.com/cms-sw/cmssw/pull/29141 uses mostly-sleeping threads to achieve "pseudo asynchronous" behavior with RPC API's that do not provide fully asynchronous API (thread does the remote calls, then blocks the thread to wait for response, and after waking up notifies the `edm::WaitingTaskWithArenaHolder` that the work is done). If the use such threads starts to spread (more classes, more modules), it could be beneficial to abstract that behavior.

In a sense `std::async()` does the job, except
* thread behavior is unspecified (create+join every time vs pool)
* returns a `future` with which one can either query the state, or block the calling thread
  * Concurrency TS adds continuation passing which is the behavior we want

One option would be to create `edm::async()` that uses a specific thread pool to minimize thread creation and joining, and continuation passing (e.g. requiring `edm::WaitingTaskWithArenaHolder` be passed around).

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This PR adds `edm::async()` facility described in https://github.com/cms-sw/cmssw/issues/29188, but implemented as a `edm::Async` service base class and `AsyncService` implementation (rationale for service-based approach instead of a free function can be found from https://github.com/cms-sw/cmssw/pull/44901#issuecomment-2103335066). This PR also replaces the use of `cudaStreamAddCallback()` with `edm::async()` accompanied with `cudaEventSynchronize()`, and makes the CUDA/Alpaka events to be created with `cudaEventBlockingSync` flag.

Measurements that I showed in CHEP 2023 https://indico.jlab.org/event/459/contributions/11810/ suggested possible 1 % throughput improvement at the HLT (of that time, many things have changed since) over `cudaStreamAddCallback()`. Earlier studies done with a prototype in https://github.com/cms-patatrack/pixeltrack-standalone/pull/321 that somehow the thread pool with `cudaEventSynchronize()` used less CPU than `cudaStreamAddCallback()`.

During the CHEP study I also tested polling with `cudaEventQuery()`, but the "waiting thread pool" approach was more performant.

Another benefit over `cudaStreamAddCallback()` is that that function "is slated for eventual deprecation and removal", and the "replacement" `cudaLaunchHostFunc()` does not call the callback function in case of an error in the associated CUDA stream.

Resolves https://github.com/cms-sw/cmssw/issues/29188
Resolves https://github.com/cms-sw/framework-team/issues/916

#### PR validation:

Unit tests in `FWCore/Concurrency`, `HeterogeneousCore/Alpaka{Core,Test}`, `HeterogeneousCore/CUDA{Utilities,Core,Test}` succeed.

**The deployment on CUDA and Alpaka modules still needs performance testing**

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

Possibly to be backported to 14_0_X.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: FWCore/Concurrency/src/async.cc [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#include "FWCore/Concurrency/interface/async.h"
+
+namespace edm::impl {
+  WaitingThread::WaitingThread() { thread_ = std::thread(&WaitingThread::threadLoop, this); }
```
Comment (fwyzard) [inline FWCore/Concurrency/src/async.cc:4]: Could you set a meaningful thread name, so it's easier to identify this thread pool in a GDB trace ?

For example:
```suggestion
  WaitingThread::WaitingThread() {
    thread_ = std::thread(&WaitingThread::threadLoop, this);
    pthread_setname_np(thread_.native_handle(), "edm async pool");
  }
```

Or even something more elaborate with a `static constexpr` name, and a check that its length is 15 or less.

Comment (makortel) [reply FWCore/Concurrency/src/async.cc]: Thanks for the suggestion, I added the name. While verifying the behavior by running the unit test in `gdb` I discovered many threads (~8) with the `edm async pool` name. Further investigation showed that the test ended up creating two `WaitingThread`s (which I can believe), but the use of `global_control` to set allowed parallelism to 1 lead the call to `onetbb::task_arena::enqueue()` in `edm::WaitingTaskWithArenaHolder::doneWaiting()` to create a new TBB-controlled thread that inherited the `edm async pool` name (and all subsequent TBB threads created by that thread inherited the name too).

When I set the allowed parallelism to 2 (which is now done in the test), I saw 2 threads with the `edm async pool` name.

While nearly all production jobs are configured to use multiple threads, maybe it would be time to look more into trying to give names to the TBB threads etc. I'll open a separate issue on that.

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
I've compared some different approaches using the current HLT menu and recent data.

### `alpaka::HostOnlyTask`
Using `CMSSW_14_0_6_MULTIARCHS` as the baseline, without any changes, I get:
```
Running 4 times over 10300 events with 8 jobs, each with 32 threads, 24 streams and 1 GPUs
   692.8 ±   0.2 ev/s (10000 events, 98.2% overlap)
   691.2 ±   0.2 ev/s (10000 events, 98.3% overlap)
   692.4 ±   0.2 ev/s (10000 events, 98.2% overlap)
   690.7 ±   0.2 ev/s (10000 events, 98.9% overlap)
 --------------------
   691.8 ±   1.0 ev/s
```

### alpaka native host tasks
Using  `CMSSW_14_0_6_MULTIARCHS` with a modified `HeterogeneousCore/AlpakaCore/src/alpaka/EDMetadata.cc`:
```c++
  void EDMetadata::enqueueCallback(edm::WaitingTaskWithArenaHolder holder) {
    alpaka::enqueue(*queue_, [holder = std::move(holder)]() {
      // The functor is required to be const, but the original waitingTaskHolder_
      // needs to be notified...
      std::exception_ptr eptr;
      const_cast<edm::WaitingTaskWithArenaHolder&>(holder).doneWaiting(eptr);
    });
  }
```
I get:
```
Running 4 times over 10300 events with 8 jobs, each with 32 threads, 24 streams and 1 GPUs
   671.0 ±   0.2 ev/s (10000 events, 98.1% overlap)
   668.1 ±   0.2 ev/s (10000 events, 98.3% overlap)
   673.5 ±   0.2 ev/s (10000 events, 99.2% overlap)
   671.9 ±   0.2 ev/s (10000 events, 97.3% overlap)
 --------------------
   671.1 ±   2.3 ev/s
```

### `edm::async`
Using `CMSSW_14_0_6_MULTIARCHS` with the changes from this PR (as of f8322a08368fca17ce3c3ab9b58fecee94b1e1c0):
```
Running 4 times over 10300 events with 8 jobs, each with 32 threads, 24 streams and 1 GPUs
   703.8 ±   0.2 ev/s (10000 events, 97.2% overlap)
   703.5 ±   0.2 ev/s (10000 events, 98.8% overlap)
   700.3 ±   0.2 ev/s (10000 events, 98.4% overlap)
   701.1 ±   0.2 ev/s (10000 events, 99.0% overlap)
 --------------------
   702.2 ±   1.8 ev/s
```

---

Bottom line, this PR speeds up the HLT by about 1.5% ± 0.3% .

**Comment by fwyzard** [comment]
+heterogeneous

**Comment by makortel** [comment]
Notes from review discussion with @Dr15Jones 
- Have `edm::async()` to take the `WaitingThreadPool` as an argument
- Move the ownership of `WaitingThreadPool` object from a function-static (i.e. global) to a Service (that would be part of the default `cmsRun` services)
- Create a base class for the Service in order to decouple things. The base class would be placed in `FWCore/Concurrency`, and the Service implementation in `FWCore/Services`
- The Service base class would have a member function with similar interface as the `edm::async()` in this PR presently
- Additional benefits/opportunities of the Service approach
  - We can add ActivityRegistry signals around the caller-supplied functor (beyond this PR)
  - We can add caller-supplied context information in the exception message from the functor

Comments beyond this PR
- We should add a test where an exception (from the external worker) is passed to the `WaitingTaskWithArenaHolder::doneWaiting()`, and see what context is in the message

File: FWCore/Concurrency/src/async.cc [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    } catch (...) {
+    }
+  }
+
+  void WaitingThread::threadLoop() {
+    std::unique_lock lk(mutex_);
+
+    do {
+      cond_.wait(lk, [this]() { return static_cast<bool>(func_) or stopThread_; });
+      if (func_) {
```
Comment (makortel) [inline FWCore/Concurrency/src/async.cc:43]: Check and act on `stopThread_` first (to avoid really strange failures)

File: FWCore/Concurrency/src/async.cc [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    CMS_SA_ALLOW try {
+      if (not stopThread_) {
+        stopThread();
+      }
+      thread_.join();
+    } catch (...) {
+    }
+  }
+
+  void WaitingThread::threadLoop() {
```
Comment (makortel) [inline FWCore/Concurrency/src/async.cc:38]: Mark this function as `noexcept`

File: FWCore/Concurrency/src/async.cc [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  WaitingThread::WaitingThread() {
+    thread_ = std::thread(&WaitingThread::threadLoop, this);
+    static constexpr auto poolName = "edm async pool";
+    // pthread_setname_np() string length is limited to 16 characters,
+    // including the null termination
+    static_assert(std::string_view(poolName).size() < 16);
+
+    int err = pthread_setname_np(thread_.native_handle(), poolName);
+    // According to the glibc documentation, the only error
+    // pthread_setname_np() can returkn is about the argument C-string
```
Comment (makortel) [inline FWCore/Concurrency/src/async.cc:18]:
```suggestion
    // pthread_setname_np() can return is about the argument C-string
```

File: FWCore/Concurrency/src/async.cc [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    // if the limit shortens, or other error conditions get added,
+    // let's assert() anyway (exception feels overkill)
+    assert(err == 0);
+  }
+
+  WaitingThread::~WaitingThread() noexcept {
+    // When we are shutting down, we don't care about any possible
+    // system errors anymore
+    CMS_SA_ALLOW try {
+      if (not stopThread_) {
```
Comment (makortel) [inline FWCore/Concurrency/src/async.cc:30]: Remove the if, and make `stopThread()` function private.

**Comment by makortel** [comment]
I addressed the comments in https://github.com/cms-sw/cmssw/pull/44901#issuecomment-2103335066 except for
> * We can add ActivityRegistry signals around the caller-supplied functor (beyond this PR)
> * We can limit the async launching to specific parts of the data processing (like only in module `acquire()` functions)

that I decided to leave for later PRs (and not to be backported to 14_0_X, unless requested).

Starting with this PR @Dr15Jones and I decided to start collecting generally useful testing modules to `FWCore/TestModules` (later on we are planning to move other modules there e.g. from `FWCore/Framework/test/stubs` and `FWCore/Integration`).

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
@Dr15Jones Please review

File: FWCore/TestModules/plugins/StreamIDFilter.cc [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+  void StreamIDFilter::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+    desc.add<std::vector<unsigned int>>("rejectStreams")
+        ->setComment("Stream IDs for which to reject events. If empty, all events are accepted.");
+    descriptions.addDefault(desc);
+  }
+}  // namespace edmtest
+
+DEFINE_FWK_MODULE(edmtest::StreamIDFilter);
```
Comment (makortel) [inline FWCore/TestModules/plugins/StreamIDFilter.cc:35]: @Dr15Jones What convention do we want here? I can think of
* `edmtest::<module>`
* `edm::<module>`
* Have the `<module>` defined in some namespace, but define it to the framework via `<module>` type alias (i.e. without a namespace)

I started with the `edmtest::<module>`, but at this stage it would be easy to change.

Comment (makortel) [reply FWCore/TestModules/plugins/StreamIDFilter.cc:35]: Or maybe other way to phrase the question would be if `edmtest` would be appropriate namespace for such generally-usable modules, or do we want to preserve it to modules that are specific for testing the framework?

**Comment by makortel** [comment]
On another look, I think I'm still missing

> * We can destruct the `WaitingThreadPool` at the endJob, or Service destructor, instead of "after `main()`"

I'll look into it.

File: FWCore/Concurrency/interface/WaitingThreadPool.h [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+      WaitingThread(WaitingThread const&) = delete;
+      WaitingThread& operator=(WaitingThread&&) = delete;
+      WaitingThread(WaitingThread&&) = delete;
+      WaitingThread& operator=(WaitingThread const&) = delete;
+
+      template <typename F, typename G>
+      void run(WaitingTaskWithArenaHolder holder,
+               F&& func,
+               G&& errorContextFunc,
+               std::shared_ptr<WaitingThread> thisPtr) {
```
Comment (Dr15Jones) [inline FWCore/Concurrency/interface/WaitingThreadPool.h:28]: Should this be `noexcept`?

File: FWCore/Concurrency/interface/WaitingThreadPool.h [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+            e.addContext(errorContext());
+            holder.doneWaiting(std::current_exception());
+          }
+        };
+        thisPtr_ = std::move(thisPtr);
+        cond_.notify_one();
+      }
+
+    private:
+      void stopThread() {
```
Comment (Dr15Jones) [inline FWCore/Concurrency/interface/WaitingThreadPool.h:45]: Could be `noexcept`

File: FWCore/Concurrency/interface/Async.h [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "FWCore/Concurrency/interface/WaitingThreadPool.h"
+
+namespace edm {
+  // All member functions are thread safe
+  class Async {
+  public:
+    Async() = default;
+    virtual ~Async();
```
Comment (Dr15Jones) [inline FWCore/Concurrency/interface/Async.h:12]: Should this be marked `noexcept`?

File: FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+    std::atomic<int> count{0};
+
+    oneapi::tbb::task_group group;
+    edm::FinalWaitingTask waitTask{group};
+
+    {
+      using namespace edm::waiting_task::chain;
+      auto h = first([&pool, &count](edm::WaitingTaskHolder h) {
+                 edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                 pool.runAsync(
```
Comment (Dr15Jones) [inline FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:52]: What is the underlying behavior you are trying to test here?

File: FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc [hunk 11]
```diff
@@ -0,0 +1,258 @@
...
+                      h2, [&count]() { ++count; }, errorContext);
+                }) |
+                lastTask(edm::WaitingTaskHolder(group, &waitTask));
+
+      auto h2 = first([&pool, &count](edm::WaitingTaskHolder h) {
+                  edm::WaitingTaskWithArenaHolder h2(std::move(h));
+                  pool.runAsync(
+                      h2, [&count]() { ++count; }, errorContext);
+                }) |
+                lastTask(edm::WaitingTaskHolder(group, &waitTask));
```
Comment (Dr15Jones) [inline FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:86]: How do you mean this test to be different from the one above?

File: FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+                         }
+                         ++count;
+                       },
+                       errorContext);
+                 }) |
+                 lastTask(edm::WaitingTaskHolder(group, &waitTask));
+        h.doneWaiting(std::exception_ptr());
+      }
+      REQUIRE_THROWS_AS(waitTask.wait(), cms::Exception);
+      REQUIRE(count.load() == 1);
```
Comment (Dr15Jones) [inline FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:152]: Is this actually guaranteed to work? Couldn't `"1"` thread start and pause at line 134 then `"2"` thread finishes and only after that `"1"` continues? In that case `count.load() == 2`.

File: FWCore/Services/plugins/AsyncService.cc [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+  private:
+    void ensureAllowed() const final;
+
+    std::atomic<bool> allowed_ = true;
+  };
+
+  AsyncService::AsyncService(ParameterSet const& iConfig, ActivityRegistry& iRegistry) {
+    iRegistry.watchPreSourceEarlyTermination([this](TerminationOrigin) { allowed_ = false; });
+    iRegistry.watchPreGlobalEarlyTermination([this](GlobalContext const&, TerminationOrigin) { allowed_ = false; });
+    iRegistry.watchPreStreamEarlyTermination([this](StreamContext const&, TerminationOrigin) { allowed_ = false; });
```
Comment (Dr15Jones) [inline FWCore/Services/plugins/AsyncService.cc:25]: Should we also add `postEndJob`?

File: FWCore/TestModules/plugins/StreamIDFilter.cc (hunk 6 above)
Comment (Dr15Jones) [reply FWCore/TestModules/plugins/StreamIDFilter.cc:35]: I think we want to use the `edmtest` for any useful module that is meant for testing and NOT meant for use in production. I also think `edmtest` should appear in the python configurations as it clearly shows this is a test.

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

**Comment by makortel** [comment]
This update should fix the failed unit tests (last part of `TestFWCoreServicesDriver_AsyncService` turned out to be flawed, so that lead to a big larger reorganization)

**Comment by makortel** [comment]
@cmsbuild, please test

To see the tests actually got fixed before proceeding to @Dr15Jones' comments

File: FWCore/TestModules/plugins/StreamIDFilter.cc (hunk 6 above)
Comment (makortel) [reply FWCore/TestModules/plugins/StreamIDFilter.cc:35]: Thanks, so we can continue with my present approach.

File: FWCore/Concurrency/interface/Async.h [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+
```
[the remark this answers is not in the document]
Comment (makortel) [reply FWCore/Concurrency/interface/Async.h:13]: The `WaitingThreadPool` member already prohibits copying and moving, but it's probably good to be explicit here.

File: FWCore/Concurrency/src/WaitingThreadPool.cc [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+    }
+  }
+
+  void WaitingThread::threadLoop() noexcept {
+    std::unique_lock lk(mutex_);
+
+    while (true) {
+      cond_.wait(lk, [this]() { return static_cast<bool>(func_) or stopThread_; });
+      if (stopThread_) {
+        break;
```
[the remark this answers is not in the document]
Comment (makortel) [reply FWCore/Concurrency/src/WaitingThreadPool.cc:42]: The question came to my mind as well, so I'll add the assert.

File: FWCore/Concurrency/interface/WaitingThreadPool.h (hunk 7 above)
Comment (makortel) [reply FWCore/Concurrency/interface/WaitingThreadPool.h:28]: Depends if we want possible system errors from locking the mutex to be reported via exceptions to be caught by the caller, or via `std::terminate()`.

File: FWCore/Concurrency/interface/Async.h (hunk 9 above)
Comment (makortel) [reply FWCore/Concurrency/interface/Async.h:12]: It would probably be good. Looking what gets done by the destructor, presently the destructor of `edm::ReusableObjectHolder` is not `noexcept`. Could or should it be?
 https://github.com/cms-sw/cmssw/blob/774f85421329c42cd1d30ec43d470f527141fb92/FWCore/Utilities/interface/ReusableObjectHolder.h#L91-L97
The OneAPI spec doesn't say anything about exceptions in `oneapi::tbb::concurrent_queue<...>::try_pop()` https://spec.oneapi.io/versions/latest/elements/oneTBB/source/containers/concurrent_queue_cls/safe_member_functions.html#popping-elements
so I guess the question is what we want the behavior to be if the destructor of `T` throws.

Well, if the destructor of `T` would throw, the `while` loop would terminate early and leave the remaining `T` to not be destructed, which is probably bad enough to just call `std::terminate` (i.e. suggesting to mark the `ReusableObjectHolder` destructor as `noexcept`.

File: FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc (hunk 10 above)
Comment (makortel) [reply FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:52]: The system works also when there are multiple `runAsync()` calls from the same task (i.e. both functions have finished at the time `FinalWaitingTask::waitNoThrow()` returns).

File: FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc (hunk 11 above)
Comment (makortel) [reply FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:86]: My intention here is to have `runAsync()` called concurrently (from two different tasks).

File: FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc (hunk 12 above)
Comment (makortel) [reply FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:152]: Good point, thanks. I need to use `fetch_add()`.

File: FWCore/Services/plugins/AsyncService.cc (hunk 13 above)
Comment (makortel) [reply FWCore/Services/plugins/AsyncService.cc:25]: I see it would be consistent. I could do it already in `preEndJob`, but then the exception message should be improved, and in a way that would be step towards restricting more when the `AsyncService` can be used that I decided to leave for a later PR. I'll add the `postEndJob`.

File: FWCore/Concurrency/interface/WaitingThreadPool.h (hunk 7 above)
Comment (makortel) [reply FWCore/Concurrency/interface/WaitingThreadPool.h:28]: Let's leave as it is.

File: FWCore/Concurrency/interface/WaitingThreadPool.h (hunk 8 above)
Comment (makortel) [reply FWCore/Concurrency/interface/WaitingThreadPool.h:45]: Let's leave as it is.

File: FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc (hunk 11 above)
Comment (makortel) [reply FWCore/Concurrency/test/test_catch2_WaitingThreadPool.cc:86]: Add spinloops such that there would be an attempt to call `pool.runAsync()` simultaneously.

**Comment by makortel** [comment]
Addressed the review comments from @Dr15Jones

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by makortel** [comment]
Unit test fails in IBs https://github.com/cms-sw/cmssw/issues/45101

**Comment by makortel** [comment]
+core

**Comment by makortel** [comment]
+heterogeneous

Based on earlier https://github.com/cms-sw/cmssw/pull/44901#issuecomment-2099490033, and that on `heterogeneous` side there were no substantial changes.

**Comment by makortel** [comment]
ignore tests-rejected with ib-failure

**Comment by rappoccio** [comment]
+1

**Comment by rappoccio** [comment]
merge
